src/V3Delayed.cpp

```diff
@@ -421,39 +473,178 @@ class DelayedVisitor final : public VNVisitor {
-            ++m_statSharedSet;
-        } else {
-            // Create new flag
-            m_setVscp = createNewVarScope(vscp, "__VdlySet" + record.suffix, 1);
-            // Set it here
-            insert(new AstAssign{flp, new AstVarRef{flp, m_setVscp, VAccess::WRITE},
-                                 new AstConst{flp, AstConst::BitTrue{}}});
-            // Add the 'Pre' ordered reset for the flag
-            activep->addStmtsp(new AstAssignPre{flp, new AstVarRef{flp, m_setVscp, VAccess::WRITE},
-                                                new AstConst{flp, 0}});
-        };
+        if (m_vscpAux(vscp).needsCommitQueue) {
+            if (FileLine* const badFlp = m_vscpAux(vscp).nonDeferredFlp) {
+                insertionPointp->v3warn(
+                    BLKLOOPINIT,
+                    "Unsupported: Non-blocking assignment to array in both "
+                    "loop and suspendable process/fork\n"
+                        << badFlp->warnOther()
+                        << "... Location of non-blocking assignment in suspendable process/fork\n"
+                        << badFlp->warnContextSecondary() << insertionPointp->warnOther()
+                        << "... Location of non-blocking assignment inside loop\n");
+                return;
+            }
 
-        // Create/Get 'if (__VdlySet) { ... commit .. }'
-        AstIf* ifp = nullptr;
-        if (postp->user2p() == m_setVscp) {
-            // Optimize as above. If sharing VdlySet *ON SAME VARIABLE*, we can share the 'if'
-            ifp = VN_AS(postp->user1p(), If);
+            // Need special handling for variables that require partial updates
+            const bool partial = m_vscpAux(vscp).hasPartialUpdate;
+
+            // If partial updates are required, construct the mask and the widened value
+            AstNodeExpr* valuep = record.rhsp;
+            AstNodeExpr* maskp = nullptr;
+            if (partial) {
+                // Type of array element
+                AstNodeDType* const eDType = [&]() -> AstNodeDType* {
+                    AstNodeDType* dtypep = vscp->dtypep();
+                    while (AstUnpackArrayDType* const ap = VN_CAST(dtypep, UnpackArrayDType)) {
+                        dtypep = ap->subDTypep();
+                    }
+                    return dtypep;
+                }();
+
+                if (AstNodeExpr* const sLsbp = lhsComponents.selLsbp) {
+                    // This is a partial assignment. Need to create a mask and widen the value to
+                    // element size.
+                    AstConst* const sWidthp = lhsComponents.selWidthp;
+
+                    // Create a temporary variable with the given name, and with 'eDType' type,
+                    // with the bits selected by 'sLsbp'/'sWidthp' set to 'insertp'. Returns a
+                    // read reference to the temporary variable.
+                    auto createWidened = [&](const std::string& name, AstNodeExpr* insertp) {
```

Comment:
Please use functions -- generally much easier to understand data flow and what's being passed in with functions - I can see any m_ are from the class, and rest is passed. Otherwide needi "wider" understanding of several screen-fulls of code.

Reply:
Done, though for the record I strongly disagree with almost all that reasoning, but let's save that conversation for some other time ;)